### PR TITLE
Add tag endpoints

### PR DESCRIPTION
**Pull request by AlberdaNazarius**
- added tag endpoints
- fixed bug with returning null. 
- added asyn

### CODE REVIEW

File: Controllers/QuestionController.cs [hunk 1]
```diff
@@ -21,29 +21,47 @@ public QuestionController(IQuestionRepository questionRepository, IMapper mapper
...
 
+    [HttpPost("{questionId:long}/addTag")]
+    [ProducesResponseType(204)]
+    [ProducesResponseType(404)]
+    public async Task<IActionResult> AddTag(long questionId, [FromQuery] long tagId)
+    {
+        var question = await _questionRepository.GetQuestion(questionId);
+        
+        if (question == null)
+            return NotFound($"Question with id = {questionId} was not found");
```
Comment (RomanMykush) [inline Controllers/QuestionController.cs]: Change "=" to ":" for consistency.

File: Controllers/QuestionController.cs [hunk 2]
```diff
@@ -84,20 +102,20 @@ public IActionResult UpdateQuestion([FromBody] QuestionDto? updatedDtoObject, lo
...
     [HttpDelete("{id:long}")]
-    [ProducesResponseType(200)]
+    [ProducesResponseType(204)]
     [ProducesResponseType(400)]
     [ProducesResponseType(404)]
-    public IActionResult DeleteQuestion(long id)
+    public async Task<IActionResult> DeleteQuestion(long id)
     {
-        var questionToDelete = _questionRepository.GetQuestion(id);
+        var questionToDelete = _questionRepository.GetQuestion(id).Result;
```
Comment (RomanMykush) [inline Controllers/QuestionController.cs]: Change this to await.

File: Controllers/TagController.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        if (!ModelState.IsValid)
+            return BadRequest(ModelState);
+   
+        var objectToSave = _mapper.Map<Tag>(dto);
+        var result = await _tagRepository.AddTag(objectToSave);
+        
+        if (!result)
+            return BadRequest(ModelState);
+        
+        return Ok("Successfully created");
```
Comment (RomanMykush) [inline Controllers/TagController.cs]: Specify exactly what is being created

File: Persistence/Repositories/Impl/QuestionRepository.cs [hunk 4]
```diff
@@ -11,45 +12,79 @@ public QuestionRepository(AppDbContext context)
...
+            .Include(q => q.QuestionTags)
+            .Include(q => q.Answers)
+            .FirstOrDefaultAsync(q => q.Id == id);
+
+        if (question == null)
+            return null;
+
+        foreach (var questionTag in question.QuestionTags!)
+        {
+            var tag = await _context.Tags!.FirstOrDefaultAsync(e => e.Id == questionTag.TagId);
```
Comment (RomanMykush) [inline Persistence/Repositories/Impl/QuestionRepository.cs]: Why "e" and not "t"?
```suggestion
            var tag = await _context.Tags!.FirstOrDefaultAsync(t => t.Id == questionTag.TagId);
```

File: Persistence/Repositories/Impl/QuestionRepository.cs [hunk 5]
```diff
@@ -11,45 +12,79 @@ public QuestionRepository(AppDbContext context)
...
+            questionTag.Tag = tag;
+        }
+
+        foreach (var answer in question.Answers!)
+        {
+            var answerWithAuthor = await _context.Answers!
+                .Include(e => e.Author)
+                .FirstOrDefaultAsync(e => e.Id == answer.Id);
```
Comment (RomanMykush) [inline Persistence/Repositories/Impl/QuestionRepository.cs]: Same as previous.
```suggestion
                .Include(a => a.Author)
                .FirstOrDefaultAsync(a => a.Id == answer.Id);
```

File: Persistence/Repositories/Impl/QuestionRepository.cs [hunk 6]
```diff
@@ -11,45 +12,79 @@ public QuestionRepository(AppDbContext context)
...
     }
-
-    public bool AddQuestion(Question question, long tagId)
+    
+    public async Task<bool> AddTag(long questionId, long tagId)
     {
-        var tag = _context.Tags!.FirstOrDefault(t => t.Id == tagId);
-
+        var question = await _context.Questions!.FirstOrDefaultAsync(e => e.Id == questionId);
+        var tag = await _context.Tags!.FirstOrDefaultAsync(e => e.Id == tagId);
```
Comment (RomanMykush) [inline Persistence/Repositories/Impl/QuestionRepository.cs]: Same.
```suggestion
        var question = await _context.Questions!.FirstOrDefaultAsync(q => q.Id == questionId);
        var tag = await _context.Tags!.FirstOrDefaultAsync(t => t.Id == tagId);
```

File: Persistence/Repositories/Impl/TagRepository.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    }
+    
+    public async Task<ICollection<Tag>> GetTags()
+    {
+        return await _context.Tags!.ToListAsync();
+    }
+
+    public async Task<Tag?> GetTag(long id)
+    {
+        return await _context.Tags!.FirstOrDefaultAsync(e => e.Id == id);
```
Comment (RomanMykush) [inline Persistence/Repositories/Impl/TagRepository.cs]: Same.
```suggestion
        return await _context.Tags!.FirstOrDefaultAsync(t => t.Id == id);
```

**Review by RomanMykush — CHANGES_REQUESTED** [review]
Change Task<IActionResult> in contollers to Task<ActionResult<some_type>>. Left as is if no value will be returned. Also complete those other minor requests.

File: Controllers/QuestionController.cs (hunk 1 above)
Comment (AlberdaNazarius) [reply Controllers/QuestionController.cs:51]: > Change "=" to ":" for consistency.

done

**Pull request merged**